### PR TITLE
Fix: #2853 Updated the register method signature

### LINKED ISSUE #2853: AbstractJKubePlugin: Update return type for register method
### Component

Kubernetes Gradle Plugin

### Description
register method from AbstractJKubePlugin returns the registeredTask which never gets used. (None of the registerTasks returned from this method are saved)

```java
protected final <T extends JKubeTask> TaskProvider<T> register(Project project, String name, Class<T> type) {
    final TaskProvider<T> registeredTask = project.getTasks().register(name, type, extensionClass);
    registeredTasks.add(registeredTask);
    return registeredTask;
  }
``` 
#### Usage

 ```java
Class: KubernetesPlugin

@Override
  protected void jKubeApply(Project project) {
    register(project, "k8sApply", KubernetesApplyTask.class);
    register(project, "k8sBuild", KubernetesBuildTask.class);
    register(project, "k8sConfigView", KubernetesConfigViewTask.class);
    register(project, "k8sDebug", KubernetesDebugTask.class);
    register(project, "k8sLog", KubernetesLogTask.class);
    register(project, "k8sPush", KubernetesPushTask.class);
    register(project, "k8sResource", KubernetesResourceTask.class);
    register(project, "k8sUndeploy", KubernetesUndeployTask.class);
    register(project, "k8sHelm", KubernetesHelmTask.class);
    register(project, "k8sHelmPush", KubernetesHelmPushTask.class);
    register(project, "k8sHelmLint", KubernetesHelmLintTask.class);
    register(project, "k8sRemoteDev", KubernetesRemoteDevTask.class);
    register(project, "k8sWatch", KubernetesWatchTask.class);
  }
  
Class: OpenShiftPlugin
  
  @Override
  protected void jKubeApply(Project project) {
    register(project, "ocApply", OpenShiftApplyTask.class);
    register(project, "ocBuild", OpenShiftBuildTask.class);
    register(project, "ocConfigView", KubernetesConfigViewTask.class);
    register(project, "ocDebug", OpenShiftDebugTask.class);
    register(project, "ocLog", KubernetesLogTask.class);
    register(project, "ocPush", OpenShiftPushTask.class);
    register(project, "ocResource", OpenShiftResourceTask.class);
    register(project, "ocUndeploy", OpenShiftUndeployTask.class);
    register(project, "ocHelm", OpenShiftHelmTask.class);
    register(project, "ocHelmPush", OpenShiftHelmPushTask.class);
    register(project, "ocHelmLint", OpenShiftHelmLintTask.class);
    register(project, "ocRemoteDev", OpenShiftRemoteDevTask.class);
    register(project, "ocWatch", OpenShiftWatchTask.class);
  }
``` 

  

### Expected Behavior
Update the return type for the register method as void

```java
protected final <T extends JKubeTask> void register(Project project, String name, Class<T> type) {
    final TaskProvider<T> registeredTask = project.getTasks().register(name, type, extensionClass);
    registeredTasks.add(registeredTask);
  }
``` 

### Acceptance Criteria
- [ ] AC1 : Registered tasks should not be returned
- [ ] AC2 : Project builds okay after running mvn clean install

### PR DESCRIPTION
**Pull request by akshay11agrawal**
## Description

Fixes #2853

## Type of change

 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] I have read the [contributing guidelines](https://www.eclipse.dev/jkube/contributing)
 - [ ] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [x] My code follows the style guidelines of this project
 - [x] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [x] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [x] I have performed a self-review of my code
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [x] I have added tests that prove my fix is effective or that my feature works
 - [x] New and existing unit tests pass locally with my changes
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
### Eclipse JKube [CI Report](https://github.com/jkubeio/ci)

Started new GH workflow run for https://github.com/eclipse/jkube/pull/2863 (_2024-03-29T00:26:25Z_)

:gear: [JKube E2E Tests (8475404280)](https://github.com/jkubeio/ci/actions/runs/8475404280) :hourglass_flowing_sand:
- :hourglass_flowing_sand: [Bootstrap CI test run (#2863)](https://github.com/jkubeio/ci/actions/runs/8475404280/job/23223453594)

**Comment by rohanKanojia** [comment]
@akshay11agrawal : Could you please try pushing again? Build seems to have failed to download some maven artifacts.

File: gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java [hunk 1]
```diff
@@ -90,4 +92,15 @@ void getTaskPrecedence_withValidProject_shouldReturnTaskPrecedence() {
         .containsEntry("k8sHelmPush", Collections.singletonList(KubernetesHelmTask.class))
         .containsEntry("k8sHelmLint", Collections.singletonList(KubernetesHelmTask.class));
   }
+
+  @Test
+  void register_shouldRegisterTasks_WithoutReturningTask_kubPlugin() {
+    AbstractJKubePlugin<KubernetesExtension> jkubePlugin = mock(KubernetesPlugin.class, RETURNS_DEEP_STUBS);
```
Comment (rohanKanojia) [inline gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:98]: Do we need to create mocks with deep stubbing here?

File: gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java [hunk 2]
```diff
@@ -90,4 +92,15 @@ void getTaskPrecedence_withValidProject_shouldReturnTaskPrecedence() {
...
+    doNothing().when(jkubePlugin).register(any(),any(),any());
+    // Call the register method
+    jkubePlugin.register(project, "k8sApply", KubernetesApplyTask.class);
+    // Verify that register was called with the expected arguments
```
Comment (rohanKanojia) [inline gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:102]: I don't think these comments are adding any value. What do you think?

File: gradle-plugin/openshift/src/test/java/org/eclipse/jkube/gradle/plugin/OpenShiftPluginTest.java [hunk 3]
```diff
@@ -48,4 +55,15 @@ void configurePrecedence_withValidProject_shouldReturnTaskPrecedence() {
       .containsEntry("ocHelmPush", Arrays.asList(KubernetesHelmTask.class, OpenShiftHelmTask.class))
       .containsEntry("ocHelmLint", Arrays.asList(KubernetesHelmTask.class, OpenShiftHelmTask.class));
   }
+
+  @Test
+  void register_shouldRegisterTasks_WithoutReturningTask_openShiftPlugin() {
+    Project project = mock(Project.class, RETURNS_DEEP_STUBS);
+    AbstractJKubePlugin<OpenShiftExtension> jkubePlugin = mock(OpenShiftPlugin.class, RETURNS_DEEP_STUBS);
```
Comment (rohanKanojia) [inline gradle-plugin/openshift/src/test/java/org/eclipse/jkube/gradle/plugin/OpenShiftPluginTest.java:62]: https://github.com/eclipse/jkube/pull/2863/files#r1546354087

File: gradle-plugin/openshift/src/test/java/org/eclipse/jkube/gradle/plugin/OpenShiftPluginTest.java [hunk 4]
```diff
@@ -48,4 +55,15 @@ void configurePrecedence_withValidProject_shouldReturnTaskPrecedence() {
...
+    doNothing().when(jkubePlugin).register(any(),any(),any());
+    // Call the register method
+    jkubePlugin.register(project, "ocApply", OpenShiftResourceTask.class);
+    // Verify that register was called with the expected arguments
```
Comment (rohanKanojia) [inline gradle-plugin/openshift/src/test/java/org/eclipse/jkube/gradle/plugin/OpenShiftPluginTest.java:66]: https://github.com/eclipse/jkube/pull/2863/files#r1546355792

File: gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java (hunk 2 above)
Comment (akshay11agrawal) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:102]: Agree, will remove the comments

File: gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java (hunk 1 above)
Comment (akshay11agrawal) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:98]: You are correct, that's not needed, will do the needful

**Comment by akshay11agrawal** [comment]
@rohanKanojia  please let me know if the changes look good now

**Review by rohanKanojia — APPROVED** [review]

**Comment by rohanKanojia** [comment]
@akshay11agrawal : Thanks a lot!

**Comment by akshay11agrawal** [comment]
Hi @rohanKanojia/ @manusa, are we good to merge this PR ?

**Comment by rohanKanojia** [comment]
We usually require 2 approvals in order to merge a pull request. Marc is on vacation, he would be back next week.

File: gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java [hunk 5]
```diff
@@ -90,4 +92,15 @@ void getTaskPrecedence_withValidProject_shouldReturnTaskPrecedence() {
         .containsEntry("k8sHelmPush", Collections.singletonList(KubernetesHelmTask.class))
         .containsEntry("k8sHelmLint", Collections.singletonList(KubernetesHelmTask.class));
   }
+
+  @Test
```
Comment (sunix) [inline gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:96]: Hello,
Thanks for the contribution !
I don't understand what is the purpose of these tests. Are these verifying that the method that is called before is called ? Could you clarify ? thanks

Comment (akshay11agrawal) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:96]: We have updated the method signature as part of current changes, and we need to make sure that the **register** method is correctly registering the tasks, even though now it won't return the task. ( return type changed to **void**)

In these tests, we have made sure when the **register** method is called on any AbstractJKubePlugin implementations ( KubernetesPlugin OR OpenShiftPlugin), it does not return anything and still register the task correctly.

Comment (rohanKanojia) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:96]: Sun is referring to mocking and verifying the object being tested (which looks quite odd). You shouldn't be mocking the object you're testing.

Comment (akshay11agrawal) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:95]: Yes, my bad. Thanks for pointing it out. I have updated the test cases, please have a look once.

Comment (sunix) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:95]: sorry I don't understand
```
    jkubePlugin.register(project, "k8sApply", KubernetesApplyTask.class);
    verify(jkubePlugin, times(1)).register(project,"k8sApply", KubernetesApplyTask.class);
```
So this is verifying that `jkubePlugin.register` is called, but it is just done 1 line above. I don't see the point. To me you should remove these tests.

Comment (sunix) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java:95]: I still think this is over testing. IMHO these tests are not required.

Comment (sunix) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java]: I can still see modifications in test files (import and new line added)

Comment (akshay11agrawal) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java]: Fixed

**Review by sunix — APPROVED** [review]

File: gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java (hunk 5 above)
Comment (sunix) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/KubernetesPluginTest.java]: Thanks !!!

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**